drivers: pinctrl: add pinctrl driver for ARC emsdp

SiyuanCheng-CN opened the pull request:
Add Synopsys ARC EMSDP board Pin controller for its Pmod and Arduino shield interface.

evgeniy-paltsev commented on include/zephyr/dt-bindings/pinctrl/emsdp-pinctrl.h:
```diff
@@ -0,0 +1,164 @@
+
+/**
+ * PMOD C Multiplexor
+ */
+#define PM_C_CFG0_GPIO	 ((0) << MUX_SEL4_OFFSET)
+#define PM_C_CFG0_I2C	 ((1) << MUX_SEL4_OFFSET) /* io_i2c_mst2 */
+#define PM_C_CFG0_SPI	 ((2) << MUX_SEL4_OFFSET) /* io_spi_mst1, cs_2 */
+#define PM_C_CFG0_UART3a ((3) << MUX_SEL4_OFFSET) /* io_uart3 */
+#define PM_C_CFG0_UART3b ((4) << MUX_SEL4_OFFSET) /* io_uart3 */
+#define PM_C_CFG0_PWM1	 ((5) << MUX_SEL4_OFFSET)
+#define PM_C_CFG0_PWM2	 ((6) << MUX_SEL4_OFFSET)
+
+#define PM_C_CFG1_GPIO ((0) << MUX_SEL5_OFFSET)
+
+/**
+ * Arduino Multiplexor
+ */
+#define ARDUINO_CFG0_GPIO ((0) << MUX_SEL0_OFFSET)
+#define ARDUINO_CFG0_UART ((1) << MUX_SEL0_OFFSET) /* io_uart0 */
+
+#define ARDUINO_CFG1_GPIO ((0) << MUX_SEL1_OFFSET)
+#define ARDUINO_CFG1_PWM  ((1) << MUX_SEL1_OFFSET)
+
+#define ARDUINO_CFG2_GPIO ((0) << MUX_SEL2_OFFSET)
+#define ARDUINO_CFG2_PWM  ((1) << MUX_SEL2_OFFSET)
+
+#define ARDUINO_CFG3_GPIO ((0) << MUX_SEL3_OFFSET)
+#define ARDUINO_CFG3_PWM  ((1) << MUX_SEL3_OFFSET)
+
+#define ARDUINO_CFG4_GPIO ((0) << MUX_SEL4_OFFSET)
+#define ARDUINO_CFG4_PWM  ((1) << MUX_SEL4_OFFSET)
+
+#define ARDUINO_CFG5_GPIO ((0) << MUX_SEL5_OFFSET)
+#define ARDUINO_CFG5_SPI  ((1) << MUX_SEL5_OFFSET) /* io_spi_mst0, cs_0 */
+#define ARDUINO_CFG5_PWM1 ((2) << MUX_SEL5_OFFSET)
+#define ARDUINO_CFG5_PWM2 ((3) << MUX_SEL5_OFFSET)
+#define ARDUINO_CFG5_PWM3 ((4) << MUX_SEL5_OFFSET)
+
+#define ARDUINO_CFG6_GPIO ((0) << MUX_SEL6_OFFSET)
+#define ARDUINO_CFG6_I2C  ((1) << MUX_SEL6_OFFSET) /* io_i2c_mst1 */
```
These are internal defines and we shouldn't expose them. We can move them to the `drivers/pinctrl/pinctrl_emsdp.c` itself (as we didn't use them anywhere except pinctrl_emsdp.c) or create internal header in `drivers/pinctrl/`

gmarull commented:
ping me when ready for review

Commit:
drivers: pinctrl: add pinctrl driver for ARC emsdp

Add Synopsys ARC EMSDP board Pin controller for its Pmod
and Arduino shield interface.

Signed-off-by: Siyuan Cheng <siyuanc@synopsys.com>

SiyuanCheng-CN commented:
Now it's ready for having a review, thanks @gmarull @evgeniy-paltsev @abrodkin @ruuddw

ruuddw commented on boards/arc/emsdp/emsdp-pinctrl.dtsi:
```diff
@@ -0,0 +1,124 @@
+	};
+
+	/* ARDUINO_PIN_1 */
+	arduino1_gpio: arduino1_gpio {
```
The pinmux has some constraints, not all pins can be individually set. I looks like it was decided to only explicitly list the highest number pin for each group that is programmed together?
That would at least require some explanation (documentation is lacking), or better use the name of the group instead of the individual pins (e.g. arduino_IO0_IO1).

SiyuanCheng-CN replied:
updated in doc and renamed to arduino_CFG_xxx

ruuddw commented on drivers/pinctrl/pinctrl_emsdp.c:
```diff
@@ -0,0 +1,297 @@
+
+static uint32_t mux_regs;
+
+void set_pmod_mux(uint32_t val)
```
Are these helper functions really needed? why not simply do the sys_write calls in the pinctrl_emsdp_set?

ruuddw commented on include/zephyr/dt-bindings/pinctrl/emsdp-pinctrl.h:
```diff
@@ -0,0 +1,59 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_PINCTRL_EMSDP_PINCTRL_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_PINCTRL_EMSDP_PINCTRL_H_
+
+#define EMSDP_CREG_BASE		   (0xf0001000)
```
this should be obtained from the dts file, not defined here.

evgeniy-paltsev replied:
This isn't standard peripheral and it isn't likely to be reused on some other platdom - so probably it's not a big issue.

But as we are already providing address in device tree pinctrl entries (`reg = <0xf0001000 0x100>;`) let's use if from DT.

And anyway we shouldn't have EMSDP_CREG_BASE defined in the dt-bindings header.

ruuddw reviewed (changes requested):
See commments. Also, some emsdp board documentation update to explain the pinmuxing and use would be good to add.

evgeniy-paltsev commented on drivers/pinctrl/pinctrl_emsdp.c:
```diff
@@ -0,0 +1,297 @@
+	sys_write32(val, mux_regs | PMOD_MUX_CTRL);
+}
+
+uint32_t get_pmod_mux(void)
```
All these functions should be static - we are not using them outside of this driver.

evgeniy-paltsev commented on drivers/pinctrl/pinctrl_emsdp.c:
```diff
@@ -0,0 +1,297 @@
+
+static int pinctrl_emsdp_set(uint32_t pin, uint32_t type)
+{
+	mux_regs = (EMSDP_CREG_BASE | EMSDP_CREG_PMOD_MUX_OFFSET);
```
Logically it should be (though with this exact base address and offset current expression give same result).

```suggestion
	mux_regs = (EMSDP_CREG_BASE + EMSDP_CREG_PMOD_MUX_OFFSET);
```

Same for other places. i.e: `mux_regs | ARDUINO_MUX_CTR`

evgeniy-paltsev commented on drivers/pinctrl/pinctrl_emsdp.c:
```diff
@@ -0,0 +1,289 @@
+#define ARDUINO_CFG6_GPIO ((0) << MUX_SEL6_OFFSET)
+#define ARDUINO_CFG6_I2C  ((1) << MUX_SEL6_OFFSET) /* io_i2c_mst1 */
+
+static uint32_t mux_regs;
```
We use it only in one function, so we shouldn't make it global. Let's drop it.

evgeniy-paltsev reviewed (changes requested):
Last minor changes

nashif commented:
@gmarull  please review
